lib/hallux/ordered_map.ex

```diff
@@ -0,0 +1,273 @@
+      iex> get(my_map, :b)
+      #HalluxOrderedMap<[b: "b1", b: "b2"]>
+  """
+  def get(ordered_hash_bounds_queue, key) do
+    {_, equal, _} = split(ordered_hash_bounds_queue, key)
+    equal
+  end
+
+  @doc """
+  Insert a value and key.
+
+  ## Examples
+
+      iex> my_map = new([b: "b1", a: "a", c: "c", b: "b2"])
+      #HalluxOrderedMap<[a: "a", b: "b1", b: "b2", c: "c"]>
+      iex> insert(my_map, :a, "A")
+      #HalluxOrderedMap<[a: "a", a: "A", b: "b1", b: "b2", c: "c"]>
+  """
+  def insert(ordered_hash_bounds_queue, key, value) do
+    {smaller, equal, greater} = split(ordered_hash_bounds_queue, key)
+
+    %__MODULE__{
+      finger_tree:
+        smaller.finger_tree
+        |> FingerTree.concat(equal.finger_tree)
+        |> FingerTree.snoc(OrderedHashBounds.Value.new(key, value))
+        |> FingerTree.concat(greater.finger_tree)
+    }
+  end
+
+  @doc """
+  Find all entries with a given value.
+
+  ## Examples
+
+      iex> my_map = new([b: "yes", a: "no", c: "no", d: "yes", b: "no", e: "no"])
+      #HalluxOrderedMap<[a: "no", b: "yes", b: "no", c: "no", d: "yes", e: "no"]>
+      iex> filter_value(my_map, "yes")
+      #HalluxOrderedMap<[b: "yes", d: "yes"]>
+  """
+  def filter_value(%__MODULE__{finger_tree: finger_tree}, value) do
+    %__MODULE__{
+      finger_tree: filter_value_finger_tree(finger_tree, OrderedHashBounds.hash(value), empty())
+    }
+  end
+
+  defp filter_value_finger_tree(%Empty{}, _, accum) do
+    accum
+  end
+
+  defp filter_value_finger_tree(finger_tree, hash, accum) do
+    {_smaller, greater_or_equal} =
+      FingerTree.split(
+        finger_tree,
+        fn %OrderedHashBounds{max: max} -> hash <= max end
+      )
+
+    {greater_or_equal, smaller} =
+      FingerTree.split(
+        greater_or_equal,
+        fn %OrderedHashBounds{min: min} -> hash > min end
+      )
+
+    equal = filter_value_among_geq(greater_or_equal, hash, empty())
+
+    filter_value_finger_tree(smaller, hash, FingerTree.concat(accum, equal))
+  end
+
+  defp filter_value_among_geq(%Empty{}, _, accum) do
+    accum
+  end
+
+  defp filter_value_among_geq(finger_tree, hash, accum) do
+    {equal, greater} =
+      FingerTree.split(
+        finger_tree,
+        fn %OrderedHashBounds{max: max} -> hash < max end
+      )
+
+    {_greater, equal_or_greater} =
+      FingerTree.split(
+        greater,
+        fn %OrderedHashBounds{min: min} -> hash >= min end
+      )
+
+    filter_value_among_geq(equal_or_greater, hash, FingerTree.concat(accum, equal))
+  end
+
+  defp empty() do
+    %Empty{monoid: OrderedHashBounds.new()}
+  end
+
+  @doc """
+  Return a list of all keys, in order.
+
+  ## Examples
+
+      >>> ordered_map = OrderedMap.new([a: 1, c: 3, b: 2, b: 4])
+      #HalluxOrderedMap<[a: 1, b: 2, b: 4, c: 3]>
+      >>> keys(ordered_map)
+      [:a, :b, :b, :c]
+  """
+  def keys(ordered_map) do
+    ordered_map
+    |> Enum.map(fn {key, _value} -> key end)
```

Comment:
Shouldn't duplicates be removed?

```elixir
|> Enum.dedup()
```

Reply:
I'd find that surprising:

```
> Keyword.keys([a: 1, a: 2])
[:a, :a]
```
https://hexdocs.pm/elixir/Keyword.html#keys/1

I try to follow that API to avoid surprises.

Reply:
A bit weird, but you have a point. :sake: